neqo-transport/src/connection/tests/datagram.rs

```diff
@@ -0,0 +1,243 @@
+    assert!(server.events().any(datagram));
+
+    client.process_input(out.unwrap(), now);
+    let datagram_acked = |e| matches!(e, ConnectionEvent::DatagramAcked);
+    assert!(client.events().any(datagram_acked));
+}
+
+#[test]
+fn datagram_lost() {
+    let (mut client, _) = create_and_connect_with_dgrams();
+
+    let dgram_sent = client.stats().frame_tx.datagram;
+    assert_eq!(client.send_dgram(DATA_SMALLER_THAN_MTU), Ok(false));
+    let _out = client.process_output(now()).dgram(); // This packet will be lost.
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent + 1);
+
+    // Wait for PTO
+    let now = now() + AT_LEAST_PTO;
+    let dgram_sent = client.stats().frame_tx.datagram;
+    let out = client.process_output(now).dgram();
+    assert!(out.is_some()); //PING probing
+                            // Datagram is not sent again.
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent);
+
+    let datagram_lost = |e| matches!(e, ConnectionEvent::DatagramLost);
+    assert!(client.events().any(datagram_lost));
+}
+
+#[test]
+fn datagram_sent_once() {
+    let (mut client, _) = create_and_connect_with_dgrams();
+
+    let dgram_sent = client.stats().frame_tx.datagram;
+    assert_eq!(client.send_dgram(DATA_SMALLER_THAN_MTU), Ok(false));
+    let _out = client.process_output(now()).dgram();
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent + 1);
+
+    // Call process_output again should not send any new Datagram.
+    let dgram_sent = client.stats().frame_tx.datagram;
+    assert!(client.process_output(now()).dgram().is_none());
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent);
+}
+
+#[test]
+fn dgram_no_allowed() {
+    let mut client = default_client();
+    let mut server = default_server();
+    connect_force_idle(&mut client, &mut server);
+    server.set_max_dgram_size(DATAGRAM_LEN_MTU);
+    assert_eq!(server.max_dgram_size(), Ok(DATAGRAM_LEN_MTU));
+    assert_eq!(server.send_dgram(DATA_MTU), Ok(false));
+
+    let out = server.process_output(now()).dgram().unwrap();
+    client.process_input(out, now());
+
+    assert_error(
+        &client,
+        &ConnectionError::Transport(Error::ProtocolViolation),
+    );
+}
+
+#[test]
+fn dgram_too_big() {
+    let mut client = new_client(
+        ConnectionParameters::default().max_datagram_frame_size(DATAGRAM_LEN_SMALLER_THAN_MTU),
+    );
+    let mut server = default_server();
+    connect_force_idle(&mut client, &mut server);
+    server.set_max_dgram_size(DATAGRAM_LEN_MTU);
+    assert!(DATAGRAM_LEN_MTU > DATAGRAM_LEN_SMALLER_THAN_MTU);
+    assert_eq!(server.max_dgram_size(), Ok(DATAGRAM_LEN_MTU));
+    assert_eq!(server.send_dgram(DATA_MTU), Ok(false));
+
+    let out = server.process_output(now()).dgram().unwrap();
+    client.process_input(out, now());
+
+    assert_error(
+        &client,
+        &ConnectionError::Transport(Error::ProtocolViolation),
+    );
+}
+
+#[test]
+fn overwrite_datagram() {
```

Comment:
Not sure what this test is adding.

Reply:
Test that a previous datagram is dropped if a new one is added and the old one is not sent yet.

Reply:
Thanks.  I completely missed that.  It's pretty obvious when I look again.